sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala

```diff
@@ -0,0 +1,152 @@
+
+import javax.annotation.Nullable
+
+import scala.annotation.tailrec
+
+import org.apache.spark.sql.catalyst.analysis.TypeCoercion.hasStringType
+import org.apache.spark.sql.catalyst.expressions.{ArrayJoin, BinaryExpression, Cast, Collate, ComplexTypeMergingExpression, ConcatWs, CreateArray, Expression, In, InSubquery, Substring}
+import org.apache.spark.sql.errors.QueryCompilationErrors
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{AbstractDataType, ArrayType, DataType, StringType}
+
+abstract class CollationTypeCasts extends TypeCoercionRule {
+  /**
+   * Extracts StringTypes from filtered hasStringType
+   */
+  @tailrec
+  private def extractStringType(dt: DataType): StringType = dt match {
+    case st: StringType => st
+    case ArrayType(et, _) => extractStringType(et)
+  }
+
+  /**
+   * Casts given expression to collated StringType with id equal to collationId only
+   * if expression has StringType in the first place.
+   * @param expr
+   * @param collationId
+   * @return
+   */
+  def castStringType(expr: Expression, st: StringType): Option[Expression] =
+    castStringType(expr.dataType, st).map { dt => Cast(expr, dt)}
+
+  private def castStringType(inType: AbstractDataType, castType: StringType): Option[DataType] = {
+    @Nullable val ret: DataType = inType match {
+      case st: StringType if st.collationId != castType.collationId => castType
+      case ArrayType(arrType, nullable) =>
+        castStringType(arrType, castType).map(ArrayType(_, nullable)).orNull
+      case _ => null
+    }
+    Option(ret)
+  }
+
+  /**
+   * Collates input expressions to a single collation.
+   */
+  def collateToSingleType(exprs: Seq[Expression]): Seq[Expression] = {
+    val st = getOutputCollation(exprs)
+
+    exprs.map(e => castStringType(e, st).getOrElse(e))
+  }
+
+  /**
+   * Based on the data types of the input expressions this method determines
+   * a collation type which the output will have. This function accepts Seq of
+   * any expressions, but will only be affected by collated StringTypes or
+   * complex DataTypes with collated StringTypes (e.g. ArrayType)
+   */
+  def getOutputCollation(expr: Seq[Expression]): StringType = {
+    val explicitTypes = expr.filter(_.isInstanceOf[Collate])
+      .map(_.dataType.asInstanceOf[StringType].collationId)
+      .distinct
+
+    explicitTypes.size match {
+      // We have 1 explicit collation
+      case 1 => StringType(explicitTypes.head)
+      // Multiple explicit collations occurred
+      case size if size > 1 =>
+        throw QueryCompilationErrors
+          .explicitCollationMismatchError(
+            explicitTypes.map(t => StringType(t).typeName)
+          )
+      // Only implicit or default collations present
+      case 0 =>
+        val implicitTypes = expr.map(_.dataType)
+          .filter(hasStringType)
+          .map(extractStringType)
+          .filter(dt => dt.collationId != SQLConf.get.defaultStringType.collationId)
```

Comment:
nit:
```
val implicitCollations =
  ...
  .distinct
if (implicitCollations > 1) {
  throw ...
} else {
  implicitCollations.headOption.getOrElse(SQLConf.get.defaultStringType)
}
```

Reply:
There is one problem with this. We said collations are only different if their collationId's are different. If we do this, we would need to change the line in else clause to something that does not look so nice. I have changed this structure in https://github.com/apache/spark/pull/45819, which is a followup for this PR.

Reply:
scala has `distinctBy`, we can distinct by collation id only.